HTCondor Docker Integration

bgruening opened the pull request:
The latest release of HTCondor supports the [`docker` universe](https://research.cs.wisc.edu/htcondor/manual/v8.5/2_12Docker_Universe.html) to encapsulate jobs into a containers.

This PR enables Galaxy to submit containers via the HTCondor job runner to your cluster.
An alternative approach would be [kubernetes](http://kubernetes.io/) which @pcm32 is working on in https://github.com/galaxyproject/galaxy/issues/1902.

@abdulrahmanazab: this needs proper testing. Would be great if you can look at it.

bgruening committed:
HTCondor and Docker documentation

bgruening committed:
enable HTCondor Docker integration

bgruening committed:
add new Docker Image. We need to have bash inside of Docker.

bgruening commented on test/functional/tools/catDocker.xml:
```diff
@@ -1,9 +1,9 @@
     <description>tail-to-head</description>
     <requirements>
-      <container type="docker">busybox:ubuntu-14.04</container>
+      <container type="docker">bgruening/busybox-bash:0.1</container>
```
Since we depended on `bash` we need a new default container.
The source can be found here: https://github.com/bgruening/docker-busybox-bash

bgruening replied:
Happy to move all repos to the Galaxy project.

jmchilton replied:
Let's revert the bash related changes in lieu of solving the problem of containers not having bash with https://github.com/galaxyproject/galaxy/pull/2282?

bgruening committed:
Add the ability to set the default handler and destination via environment variables.

bgruening committed:
Revert "add new Docker Image. We need to have bash inside of Docker."

This reverts commit 3ab43e38f5b2d63c300e99635234051956905f94.

bgruening commented:
@jmchilton reverted this particular commit. Thanks for looking at it!

jmchilton commented:
@bgruening this is obviously very exciting - thanks a bunch. How does condor know what volumes to mount in the Docker container? Is that configured globally - each container mounts Galaxy's datasets and index data? Is there an option to configure that on a per-job basis?

bgruening commented:
Sorry, I don't know much about the details of how this works. I implemented this as a request from our Norwegian friends (@abdulrahmanazab ...). For me this magically works and I know that condor is managing all input and outputs on their own if you specify something like `should_transfer_files = YES`. This can be set globally with this implementation. 

A few more informations are here: https://research.cs.wisc.edu/htcondor/manual/v8.5/2_12Docker_Universe.html

Changing it on a per-job basis is supported via the original condor job runner. For example you could create a destination with:

```
<destination id="encrypt_condor" runner="condor">
      <param id="transfer_input_files">file1,dir1,file2</param>
      <param id="encrypt_input_files">file1</param>
</destination>
```

But as I'm said I'm not an expert in this and @erasche probably knows more about it. I hope @abdulrahmanazab can test all the glory details in his setup at Tjenester for Sensitive Data (TSD).

abdulrahmanazab commented:
@jmchilton The typical setup for docker support in HTCondor is to mount specific volumes (with the same path) globally, e.g. ..../database, then everything under ....database will be visible to the containers with the same path as is in Galaxy

abdulrahmanazab commented:
@bgruening Sorry about the delay which is my fault. But we had to prepare another PRACE deliverable for creating a docker job runner for slurm. I called it socker, using which slurm can run and control containers with cgroups. We may create another pull request soon ;) I'm flying to Colombia now for CCGrid2016. Will start running galaxy on HTCondor once I arrive.

jmchilton commented:
Thanks for the update - this clarifies a lot. I'm eager to merge this - I have a couple small implementation improvements I'd like to make first - if I haven't opened a PR against this branch by Monday afternoon with those though I'll just merge this as is.

jmchilton committed:
Revise HTCondor Docker image resolution.

Bring inline with the rest of the job runner's Docker image resolution stuff. Order of reosolution is:

 - Use destination specified override if found: docker_container_id_override
 - Use the tool specified container if not found and no override: see tool
 - Else - use docker_default_container_id.

This adjusts the documentation to reflect this and make the the parameters more consistent throughout. I've modified this so docker_image (the old name introduced by the Docker Condor PR) still works - but I have left it undocumented intentionally for consistency.

jmchilton commented:
So my comments (implemented as a PR https://github.com/bgruening/galaxy/pull/10) are pretty much the same as my comments on the Kubernates PR (#2314) - I'm excited for the    change I just want to remain consistent in how we resolve docker identifiers throughout tools and jobs in Galaxy.

This way - the work done to decompose the docker options as part of #2314 for instance will be instantly usable by this condor runtime mode.

bgruening committed:
Merge pull request #10 from jmchilton/htcondor_docker

Revise HTCondor Docker image resolution.

bgruening commented:
@jmchilton thanks a lot for looking at it! If this all works out like planned .. we have an amazing new feature to play with :)
You rock!

bgruening commented:
Thanks for merging @jmchilton!
